### PR TITLE
[IntraNodeComm] accept P2P buffer size as constructor argument

### PR DESCRIPTION
**Pull request by yifuwang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #118038
* __->__ #120856
* #120855

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225

### CODE REVIEW

File: torch/csrc/distributed/c10d/intra_node_comm.cu [hunk 1]
```diff
@@ -636,25 +638,33 @@ at::Tensor IntraNodeComm::hybridCubeMeshAllReduce(
 
 AllReduceAlgo IntraNodeComm::selectAllReduceAlgo(const at::Tensor& input) {
   // Only support bf16 for now
-  if (input.dtype() != at::kBFloat16 ||
-      static_cast<size_t>(input.numel() * input.element_size()) >
-          kMaxIntraNodeSize) {
+  if (input.dtype() != at::kBFloat16) {
     return AllReduceAlgo::NONE;
```
Comment (wanchaol) [inline torch/csrc/distributed/c10d/intra_node_comm.cu:642]: wondering for the un-supported cases, we should fallback to nccl allreduce?

**Review by wanchaol — APPROVED** [review]
lgtm! one question only

File: torch/csrc/distributed/c10d/intra_node_comm.cu (hunk 1 above)
Comment (yifuwang) [reply torch/csrc/distributed/c10d/intra_node_comm.cu:642]: Right. Currently, `IntraNodeComm` is wrapped by `ProcessGroupNCCL`. `ProcessGroupNCCL::all_reduce` performs algo selection and uses the returned value to determine whether to use intra-node all-reduce or normal nccl all-reduce.

I'm gradually refactoring the code to reverse the relationship. We'll have p2p process group that wraps around a nccl pg as fallback. This way the integration will not be intrusive.

**Pull request closed without merging**